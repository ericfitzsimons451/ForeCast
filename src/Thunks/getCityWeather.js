import { isLoading, hasError, setCityWeather } from '../Actions'

export const getCityWeather = (url) => {
	return async (dispatch) => {
		try {
			dispatch(isLoading(true))
			const response = await fetch(url)
			if (!response.ok) {
				throw Error(response.statusText)
			}
			const data = await response.json()
			dispatch(isLoading(false))
			dispatch(setCityWeather(data))
		} catch (error) {
			dispatch(hasError(error.message))
		}
	}
}

