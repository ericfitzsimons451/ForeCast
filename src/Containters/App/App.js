import React, { Component } from 'react';
import './App.css';
import { weatherApiKey } from '../../Helpers/weatherApiKey'
import { Route } from 'react-router-dom'
import Header from '../../Components/Header/Header'
import Welcome from '../../Components/Welcome/Welcome'
import SearchForm from '../../Containters/SearchForm/SearchForm'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { getCityWeather } from '../../Thunks/getCityWeather'
import WeatherContainer from '../../Components/WeatherContainer/WeatherContainer'

export class App extends Component {

  getCityWeather = (cityName) => {
    const url = `https://api.weatherbit.io/v2.0/current?city=${cityName}&key=${weatherApiKey}`
    this.props.getCityWeather(url)
  }

  render() {
    return (
      <div className="App">
        <Header />
        <Route path='/' render={() => <Welcome />} />
        <SearchForm getCityWeather={this.getCityWeather} />
        {/* <WeatherContainer /> */}
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  getCityWeather: url => dispatch(getCityWeather(url))
})

export const mapStateToProps = (state) => ({
  isLoading: state.isLoading,
  hasError: state.hasError,
  currentWeather: state.currentWeather
})

App.propTypes = {
  getCityWeather: PropTypes.func,
  isLoading: PropTypes.bool,
  hasError: PropTypes.string,
  currentWeather: PropTypes.object
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
