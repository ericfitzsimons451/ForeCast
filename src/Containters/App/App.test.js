import React from 'react'
import { shallow } from 'enzyme'
import { App, mapStateToProps, mapDispatchToProps } from './App'
import { getCityWeather } from '../../Thunks/getCityWeather'
jest.mock('../../Thunks/getCityWeather')

describe('App', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe.only('getCityWeather', () => {
    it('should take in a city name and call props.getCityWeather with the url/cityName', () => {
      const mockCityName = 'Denver'
      const mockGetCityWeather = jest.fn()
      const wrapper = shallow(<App getCityWeather={mockGetCityWeather}/>)
      const weatherApiKey = 230232
      const mockUrl = `https://api.weatherbit.io/v2.0/current?city=${mockCityName}&key=${weatherApiKey}`

      expect(mockGetCityWeather).toHaveBeenCalledWith(mockUrl)
    })
  })
  
  describe('mapStateToProps', () => {
    it('should mapStateToProps', () => {
      const mockState = {}
      const updatedState = {
        isLoading: false,
        hasError: false,
        currentWeather: {
          city: 'Denver',
          sunrise: '11:30'
        }
      }
      const firstState = mapStateToProps(mockState)
      expect(firstState).toEqual({})
      const newState = mapStateToProps(updatedState)
      expect(newState).toEqual(updatedState)
    })
  })

  it('should dispatch getCityWeather via the props object', () => {
    const mockDispatch = jest.fn()
    const mockUrl = 'www.url.com'
    const actionToDispatch = getCityWeather(mockUrl)
    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.getCityWeather(mockUrl)
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })
})
