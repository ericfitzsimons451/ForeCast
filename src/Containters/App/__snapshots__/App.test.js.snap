// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Header />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/search"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/cities/:id"
      render={[Function]}
    />
    <Route
      component={[Function]}
    />
  </Switch>
</div>
`;
